tokenawarebandit.go

```diff
@@ -703,3 +714,77 @@ func (lb *latencyBucket) add(latency time.Duration) {
 	lb.sum += latency
 	lb.count += 1
 }
+
+const cacheLineSize = 64 // for amd64
+
+// bigWriterBucket keeps counts of latency with reduced lock contention.
+type bigWriterBucket struct {
+	cpus []bigWriterPerCPU
+}
+
+func newBigWriterBucket() bigWriterBucket {
+	cores := cpuid.CPU.LogicalCores
+	if cores == 0 {
+		// fall back to single locked bucket.
+		cores = 1
+	}
+	return bigWriterBucket{
+		cpus: make([]bigWriterPerCPU, cores),
+	}
+}
+
+// selectCPU selects which index to update.
+func (bw *bigWriterBucket) selectCPU() int {
+	id := cpuid.CPU.LogicalCPU()
+	if id == -1 {
+		// fall back to single locked bucket.
+		id = 0
+	}
+	// Make sure we don't overflow, e.g. if a CPU is hot-plugged.
+	return id % len(bw.cpus)
+}
+
+// add adds a latency sample.
+func (bw *bigWriterBucket) add(latency time.Duration) {
+	id := bw.selectCPU()
+	bw.cpus[id].mu.Lock()
+	bw.cpus[id].value.add(latency)
+	bw.cpus[id].mu.Unlock()
+}
+
+// readLocal reads the bucket state for the local CPU only.
+func (bw *bigWriterBucket) readLocal() latencyBucket {
+	id := bw.selectCPU()
+	bw.cpus[id].mu.Lock()
+	defer bw.cpus[id].mu.Unlock()
+	return bw.cpus[id].value
+}
+
+// reset reads the bucket state from all CPUs and resets the bucket to zero.
+func (bw *bigWriterBucket) reset() latencyBucket {
+	for i := 0; i < len(bw.cpus); i++ {
+		bw.cpus[i].mu.Lock()
+	}
+	var oldValue latencyBucket
+	for i := 0; i < len(bw.cpus); i++ {
+		oldValue.sum += bw.cpus[i].value.sum
+		oldValue.count += bw.cpus[i].value.count
+		bw.cpus[i].value = latencyBucket{}
+	}
+	for i := 0; i < len(bw.cpus); i++ {
+		bw.cpus[i].mu.Unlock()
+	}
```

Comment:
why do we need to lock all mutexes at once intead of:

```suggestion
	var oldValue latencyBucket
	for i := 0; i < len(bw.cpus); i++ {
		bw.cpus[i].mu.Lock()
		oldValue.sum += bw.cpus[i].value.sum
		oldValue.count += bw.cpus[i].value.count
		bw.cpus[i].value = latencyBucket{}
		bw.cpus[i].mu.Unlock()
	}
```

Reply:
I was thinking about this too, you are right, we can lock them sequentially.